### ISSUE TITLE
vmap does not ensure sharding propagation

### ISSUE THREAD

**Issue by francois-rozet** [google/jax#20683]
### Description

As discussed in #15680, some operations (including `jnp.fft.fft` and `jnp.linalg.cholesky`) do not preserve the sharding of arrays, leading to unnecessary communication between devices.

In my opinion, using `jax.vmap` should solve this issue, as we explicitly inform JAX/XLA that the operation is embarrassingly parallel and, hence, does not require communication along some axis. This is not the case unfortunately, meaning that v-mapped operations are not actually parallelized.

For example, in the following (using 4 GPUs) the sharding of `x` is not preserved/propagated to `y` and the XLA-generated code contains `%all-gather` instructions.

```python
mesh = jax.sharding.Mesh(jax.devices(), 'i')
spec = jax.sharding.PartitionSpec('i')

distributed = jax.sharding.NamedSharding(mesh, spec)

x = jax.numpy.ones((16, 1024))
x = jax.device_put(x, distributed)

y = jax.vmap(jax.numpy.fft.fft)(x)  # x.sharding != y.sharding

print(jax.jit(jax.vmap(jnp.fft.fft)).lower(x).compile().as_text())
```

```
HloModule jit_fft, is_scheduled=true, entry_computation_layout={(f32[4,1024]{1,0})->c64[16,1024]{1,0}}, allow_spmd_sharding_propagation_to_output={true}, num_partitions=4

%fused_computation (param_0: f32[4,1024]) -> c64[4,1024] {
  %param_0 = f32[4,1024]{1,0} parameter(0)
  ROOT %convert.3 = c64[4,1024]{1,0} convert(f32[4,1024]{1,0} %param_0), metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/convert_element_type[new_dtype=complex64 weak_type=False]" source_file="/tmp/ipykernel_1476911/1097094521.py" source_line=5}
}

ENTRY %main.10_spmd (param: f32[4,1024]) -> c64[16,1024] {
  %param = f32[4,1024]{1,0} parameter(0), sharding={devices=[4,1]<=[4]}
  %wrapped_convert.2 = c64[4,1024]{1,0} fusion(f32[4,1024]{1,0} %param), kind=kLoop, calls=%fused_computation, metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/convert_element_type[new_dtype=complex64 weak_type=False]" source_file="/tmp/ipykernel_1476911/1097094521.py" source_line=5}
  %all-gather-start = (c64[4,1024]{1,0}, c64[16,1024]{1,0}) all-gather-start(c64[4,1024]{1,0} %wrapped_convert.2), channel_id=1, replica_groups={{0,1,2,3}}, dimensions={0}, use_global_device_ids=true, metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1476911/1097094521.py" source_line=5}, backend_config={"is_sync":true,"no_parallel_custom_call":false}
  %all-gather-done = c64[16,1024]{1,0} all-gather-done((c64[4,1024]{1,0}, c64[16,1024]{1,0}) %all-gather-start), metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1476911/1097094521.py" source_line=5}
  ROOT %fft.2 = c64[16,1024]{1,0} fft(c64[16,1024]{1,0} %all-gather-done), fft_type=FFT, fft_length={1024}, frontend_attributes={fingerprint_before_lhs="9d3f02deb23b74ebbe5c21e63672e8a2"}, metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1476911/1097094521.py" source_line=5}
}
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.23
jaxlib: 0.4.23
numpy:  1.26.3
python: 3.9.18 | packaged by conda-forge | (main, Dec 23 2023, 16:33:10)  [GCC 12.3.0]
jax.devices (4 total, 4 local): [cuda(id=0) cuda(id=1) cuda(id=2) cuda(id=3)]
process_count: 1

$ nvidia-smi
Wed Apr 10 13:28:24 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.154.05             Driver Version: 535.154.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A100-SXM4-40GB          On  | 00000000:06:00.0 Off |                    0 |
| N/A   28C    P0              56W / 400W |  32224MiB / 40960MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM4-40GB          On  | 00000000:47:00.0 Off |                    0 |
| N/A   28C    P0              60W / 400W |  32208MiB / 40960MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   2  NVIDIA A100-SXM4-40GB          On  | 00000000:85:00.0 Off |                    0 |
| N/A   27C    P0              57W / 400W |  32208MiB / 40960MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   3  NVIDIA A100-SXM4-40GB          On  | 00000000:C7:00.0 Off |                    0 |
| N/A   29C    P0              57W / 400W |  32088MiB / 40960MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A   1476911      C   ...t/micromamba/envs/priors/bin/python    32212MiB |
|    1   N/A  N/A   1476911      C   ...t/micromamba/envs/priors/bin/python    32196MiB |
|    2   N/A  N/A   1476911      C   ...t/micromamba/envs/priors/bin/python    32196MiB |
|    3   N/A  N/A   1476911      C   ...t/micromamba/envs/priors/bin/python    32076MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by yashk2810** [comment]
Try specifying `spmd_axis_name` to `vmap`? That argument basically tells vmap on which axis of the mesh to shard the mapped away dimension. Since you only have 1 dimension try: `jax.vmap(jax.numpy.fft.fft, spmd_axis_name-'i')(x)` with a with_sharding_constraint on jnp.fft.fft

See an example here: https://github.com/google/jax/blob/f5cc272615ce2795f9133e63b7b535ec5ada7e52/tests/pjit_test.py#L753-L771

**Comment by francois-rozet** [comment]
Hello @yashk2810, I tried your recommendation and there is still communication between devices (`%all-gather`). Besides, adding `jax.lax.with_sharding_constraint` everywhere the `fft` is used (which could be deep inside a codebase) is not a viable approach in my opinion.

```python
mesh = jax.sharding.Mesh(jax.devices(), 'i')
spec = jax.sharding.PartitionSpec('i')

distributed = jax.sharding.NamedSharding(mesh, spec)

x = jax.numpy.ones((16, 1024))
x = jax.device_put(x, distributed)

def my_fft(x):
    x = jax.lax.with_sharding_constraint(x, jax.sharding.NamedSharding(mesh, jax.sharding.PartitionSpec(None)))
    x = jax.numpy.fft.fft(x)
    return x

jit_vmap_fft = jax.jit(jax.vmap(jax.numpy.fft.fft, spmd_axis_name='i'), in_shardings=distributed, out_shardings=distributed)

print(jit_vmap_fft.lower(x).compile().as_text())
```

```
HloModule jit_fft, is_scheduled=true, entry_computation_layout={(f32[4,1024]{1,0})->c64[4,1024]{1,0}}, num_partitions=4

%fused_computation (param_1: c64[16,1024], param_1.3: u32[]) -> c64[4,1024] {
  %param_1 = c64[16,1024]{1,0} parameter(0)
  %param_1.3 = u32[] parameter(1)
  %convert.5 = s32[] convert(u32[] %param_1.3), metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
  %constant_6 = s32[] constant(4), metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
  %multiply.3 = s32[] multiply(s32[] %convert.5, s32[] %constant_6), metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
  %constant_5 = s32[] constant(0), metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
  ROOT %dynamic-slice.3 = c64[4,1024]{1,0} dynamic-slice(c64[16,1024]{1,0} %param_1, s32[] %multiply.3, s32[] %constant_5), dynamic_slice_sizes={4,1024}, metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
}

%fused_computation.1 (param_0.1: f32[4,1024]) -> c64[4,1024] {
  %param_0.1 = f32[4,1024]{1,0} parameter(0)
  ROOT %convert.6 = c64[4,1024]{1,0} convert(f32[4,1024]{1,0} %param_0.1), metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/convert_element_type[new_dtype=complex64 weak_type=False]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
}

ENTRY %main.12_spmd (param: f32[4,1024]) -> c64[4,1024] {
  %param = f32[4,1024]{1,0} parameter(0), sharding={devices=[4,1]<=[4]}
  %wrapped_convert.3 = c64[4,1024]{1,0} fusion(f32[4,1024]{1,0} %param), kind=kLoop, calls=%fused_computation.1, metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/convert_element_type[new_dtype=complex64 weak_type=False]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
  %all-gather-start = (c64[4,1024]{1,0}, c64[16,1024]{1,0}) all-gather-start(c64[4,1024]{1,0} %wrapped_convert.3), channel_id=1, replica_groups={{0,1,2,3}}, dimensions={0}, use_global_device_ids=true, metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}, backend_config={"is_sync":true,"no_parallel_custom_call":false}
  %all-gather-done = c64[16,1024]{1,0} all-gather-done((c64[4,1024]{1,0}, c64[16,1024]{1,0}) %all-gather-start), metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
  %fft.2 = c64[16,1024]{1,0} fft(c64[16,1024]{1,0} %all-gather-done), fft_type=FFT, fft_length={1024}, metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
  %partition-id = u32[] partition-id(), metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
  ROOT %fusion = c64[4,1024]{1,0} fusion(c64[16,1024]{1,0} %fft.2, u32[] %partition-id), kind=kLoop, calls=%fused_computation, frontend_attributes={fingerprint_before_lhs="51bf9dff1b3f0073be7f87b72c053859"}, metadata={op_name="jit(fft)/jit(main)/vmap(jit(fft))/jit(fft)/fft[fft_type=FftType.FFT fft_lengths=(1024,)]" source_file="/tmp/ipykernel_1632380/1374808900.py" source_line=9}
}
```

**Comment by francois-rozet** [comment]
@yashk2810 In the following example, is there a way to modify the function `g`, such that it does *not* lead to communication between devices, regardless of `f`?

```python
import jax

mesh = jax.sharding.Mesh(jax.devices(), 'i')
spec = jax.sharding.PartitionSpec('i')

distributed = jax.sharding.NamedSharding(mesh, spec)

x = jax.numpy.ones((16, 1024))
x = jax.device_put(x, distributed)

def f(x):  # arbitrary function (impossible to modify)
    return jax.numpy.fft.fft(x)

@jax.jit
def g(x):  # my function (can modify)
    y = jax.vmap(f)(x)
    return y

print(g.lower(x).compile().as_text())  # should not contain %all-gather
```

**Comment by francois-rozet** [comment]
Using `jax.experimental.shard_map.shard_map` prevents device communication, but requires to use (and know) the `mesh` and `spec` objects within `g`.

```python
import jax

mesh = jax.sharding.Mesh(jax.devices(), 'i')
spec = jax.sharding.PartitionSpec('i')

distributed = jax.sharding.NamedSharding(mesh, spec)

x = jax.numpy.ones((16, 1024))
x = jax.device_put(x, distributed)

def f(x):  # arbitrary function (impossible to modify)
    return jax.numpy.fft.fft(x)

@jax.jit
def g(x):
    y = shard_map(
        f=f,
        mesh=mesh,
        in_specs=spec,
        out_specs=spec,
    )(A, x)

    return y

print(g.lower(x).compile().as_text())  # does not contain %all-gather
```
